docs/sphinx/examples/cpp/algorithms/bernstein_vazirani.cpp

```diff
@@ -0,0 +1,88 @@
+// Compile and run with:
+// ```
+// nvq++ bernstein_vazirani.cpp -o bv.x --target nvidia-mgpu
+// mpirun -np 4 ./bv.x
+// ```
+
+// This example is meant to demonstrate the cuQuantum
+// GPU-accelerated multi-GPU backend. A state vector simulation for
+// up to 32 qubits can be run on a single GPU. Larger state vectors
+// can be simulated using the multi-GPU backend.
+// The amount of resources required for the simulation doubles with
+// with each additional qubit.
+
+#include <bitset>
+#include <cudaq.h>
+#include <iostream>
+#include <random>
+
+template <int nrOfBits>
+std::bitset<nrOfBits> random_bits() {
+
+  std::bitset<nrOfBits> randomBits;
+
+  std::default_random_engine generator;
+  std::uniform_real_distribution<float> distribution(0.0, 1.0);
+
+  float randNum;
+  for (size_t i = 0; i < nrOfBits; i++) {
+    randNum = distribution(generator);
+    if (randNum < 0.5) {
+      randNum = 0;
+    } else {
+      randNum = 1;
+    }
+    randomBits.set(i, randNum);
+  }
+  return randomBits;
+}
+
+template <int nrOfBits>
+struct oracle {
+  auto operator()(std::bitset<nrOfBits> bitvector, cudaq::qspan<> qs,
+                  cudaq::qubit &aux) __qpu__ {
+
+    for (size_t i = 0; i < nrOfBits; i++) {
+      if (bitvector[i] & 1) {
+        x<cudaq::ctrl>(qs[nrOfBits - i - 1], aux);
+      }
+    }
+  }
+};
+
+template <int nrOfBits>
+struct bernstein_vazirani {
+  auto operator()(std::bitset<nrOfBits> bitvector) __qpu__ {
+
+    cudaq::qreg<nrOfBits> qs;
+    cudaq::qubit aux;
+    h(aux);
+    z(aux);
+    h(qs);
+
+    oracle<nrOfBits>{}(bitvector, qs, aux);
+    h(qs);
+    mz(qs);
+  }
+};
+
+int main() {
+  // The number of qubits can be increased when targeting
+  // the `nvidia-mgpu` backend.
+  const int nr_qubits = 32;
+  auto bitvector = random_bits<nr_qubits>();
```

Comment:
Do we need `cudaq.set_random_seed(13)` here to ensure that multiple MPI ranks are all passing the same `bitvector` into the kernel?

Reply:
No, the random number generator used sets the seed to a constant 1u by default. However, I changed the code to make this more explicit, and to allow passing in a seed as argument to main.
I am a bit confused, however, why setting the random seed for the simulator with cudaq.set_random_seed would have anything to do with the output of a random number generator in my application code?

Reply:
Sorry, you're right that `cudaq.set_random_seed()` would not help with getting the same `bitvector` into the kernel because that uses `std::default_random_engine`, and that is by far the most important one to get synchronized across the ranks. My mistake. (Thanks for making that more clear, though.) Adding `cudaq.set_random_seed` would help with repeatable results for `cudaq::sample()`. but it appears that's not important in this case because it is returning the same value 100% of the time.